### ISSUE TITLE
Use "pure-module" flag from package.json to opt modules into the more aggressive build optimization process

### ISSUE THREAD

**Issue by alexeagle** [angular/angular-cli#12103]
_From @IgorMinar on August 9, 2017 23:29_

build optimizer currently contains a whitelist for all packages that should be processed: https://github.com/angular/devkit/blob/a6cc86360ce9e8f32739995ee7961487345a3d45/packages/angular_devkit/build_optimizer/src/build-optimizer/build-optimizer.ts#L21-L33

This whitelist is for prefixing function calls with pure annotation for uglify.

Webpack is working on introducing pure module feature that npm packages opt-in to via `"pure-module": true` in package.json.

Since the meaning of this flag is the same as what we need for purification of top level function calls, we could get rid of the whitelist and use the same package.json metadata as webpack.

Since all of the modules in the whitelist will also want to benefit from the pure-modules feature in webpack, sharing the flag would allow us to get rid of the whitelist.

_Copied from original issue: angular/devkit#89_

**Comment by alexeagle** [comment]
_From @filipesilva on August 10, 2017 8:59_

For reference: https://github.com/webpack/webpack/tree/feature/pure-module/examples/pure-module

Doing this on the build-optimizer itself raises some problems. Build optimizer doesn't know about modules, and it actually doesn't know about file systems either. It only knows about file contents and (if given) file paths.

And this is relevant in a webpack context because the file contents for a given file path may or may not be related to that path on disk. For instance, the file may have been through some loaders already and it's current content changed, or the given file path may not even exist at all (this happens for AOT code).

The one that truly is aware of the `pure-module` flag is the system that is resolving the module - webpack in this case, maybe others in the future. So my suggestion is that we pass that information along in the `build-optimizer` call in the consumer:

```
const boOutput = buildOptimizer({ content: input, pure: true });
```

This way the module loader can tell build optimizer that it thinks this given module is pure, and build optimizer itself drops the whitelist and doesn't even bother checking the file path, which was a problematic way to detect these anyway.

**Comment by alexeagle** [comment]
_From @IgorMinar on August 22, 2017 1:29_

If I remember correctly from looking at the pure-module branch diff,
webpack does keep track of which modules are pure and should be able to
pass that information along to all the plugins and loaders.

**Comment by clydin** [comment]
This has been available in build optimizer since at least 6.1.0.
